src/openMVG/matching_image_collection/GPU_Matcher_Regions.cpp

```diff
@@ -0,0 +1,140 @@
+  typedef std::map<size_t, std::vector<size_t> > Map_vectorT;
+  Map_vectorT map_Pairs;
+  for (Pair_Set::const_iterator iter = pairs.begin(); iter != pairs.end(); ++iter)
+  {
+    map_Pairs[iter->first].push_back(iter->second);
+  }
+
+  // Perform matching between all the pairs
+  for (Map_vectorT::const_iterator iter = map_Pairs.begin();
+    iter != map_Pairs.end(); ++iter)
+  {
+    const size_t I = iter->first;
+    const std::vector<size_t> & indexToCompare = iter->second;
+
+    std::shared_ptr<features::Regions> regionsI = regions_provider->get(I);
+    if (regionsI->RegionCount() == 0)
+    {
+      my_progress_bar += indexToCompare.size();
+      continue;
+    }
+
+    switch (eMatcherType_) {
+      case BRUTE_FORCE_HAMMING:
+      {
+        std::vector<IndMatches> matchedPoints(indexToCompare.size());
+#ifdef OPENMVG_USE_OPENMP
+        #pragma omp parallel for schedule(dynamic)
+#endif
+        for (size_t j = 0; j < (int)indexToCompare.size(); ++j)
+        {
+          const size_t J = indexToCompare[j];
+
+          std::shared_ptr<features::Regions> regionsJ = regions_provider->get(J);
+          if (regionsJ->RegionCount() == 0
+              || regionsI->Type_id() != regionsJ->Type_id())
+          {
+#ifdef OPENMVG_USE_OPENMP
+      #pragma omp critical
+#endif
+            ++my_progress_bar;
+            continue;
+          }
+
+          switch (regionsI->DescriptorLength()) {
+            case 128: {
+#ifdef OPENMVG_USE_CUDA
+              gpu::LatchBitMatcher matcher(std::max(regionsI->RegionCount(), regionsJ->RegionCount()));
+              matcher.match(
+                const_cast<unsigned int*>(static_cast<const unsigned int*>(regionsI->DescriptorRawData())),
+                const_cast<unsigned int*>(static_cast<const unsigned int*>(regionsJ->DescriptorRawData())), 
+                regionsI->RegionCount(), 
+                regionsJ->RegionCount());
+              matchedPoints[j] = matcher.retrieveMatches(f_dist_ratio_);
+#endif
+              break;
+            }
+            default: {
+              break;
+            }
+          }
+        }
+#ifdef OPENMVG_USE_OPENMP
+        #pragma omp parallel for schedule(dynamic)
+#endif
+        for (int j = 0; j < static_cast<int>(indexToCompare.size()); j++)
+        {
+          const size_t J = indexToCompare[j];
+
+          IndMatches vec_putatives_matches;
+          for (size_t k = 0; k < matchedPoints[j].size(); k++) {
+            vec_putatives_matches.push_back(IndMatch(matchedPoints[j][k].i_, matchedPoints[j][k].j_));
+          }
+
+#ifdef OPENMVG_USE_OPENMP
+          #pragma omp critical
+#endif
+          {
+            ++my_progress_bar;
+            if (!vec_putatives_matches.empty())
+            {
+              map_PutativesMatches.insert( { {I,J}, std::move(vec_putatives_matches) });
+            }
+          }
+        }
+      }
+      default:
+      {
+        break;
```

Comment:
One step ahead of you. I've integrated the OpenCV UChar and Float GPU matchers into my personal repo. I know you'd like to keep out OpenCV from your framework, so I didn't push them. However, the programmatic interface can be easily decoupled and used here.

I would look at BruteForceHammingMatcher and BruteForceL2Matcher contained in this directory: https://github.com/mdaiter/openMVG/tree/f9aa8e98fd2e57f80d39a4720e10cae94861864c/src/openMVG/matching_image_collection/gpu